File: packages/warlog-spec-ts/src/abi.ts
/**
 * AbiConnector — abstract base class for a response-action connector
 * in TypeScript. Mirrors `warlog_spec.abi.AbiConnector` in the Python
 * package.
 *
 * A connector wraps a vendor API (CrowdStrike Falcon, SentinelOne,
 * Okta, Defender, Proofpoint, …) behind the canonical ABI lifecycle :
 *
 *     authenticate  →  dryRun  →  apply  →  verify
 *
 * The runner enforces this lifecycle. The connector implements each
 * step ; the canonical params are validated by the runner BEFORE
 * ``dryRun`` is invoked, so the connector can assume well-formed
 * input.
 *
 * Failure model : any error the connector throws is wrapped by the
 * runner into a :class:`ConnectorError`. Use :class:`ConnectorAbiError`
 * when you have an explicit failure category to surface (auth,
 * not_found, state_conflict, transient, policy) — the runner
 * preserves it verbatim. Anything else is mapped to ``TRANSIENT``
 * (retryable) by default.
 */

import type {
  ConnectorCapability,
  ConnectorError,
  FailureCategoryValue,
  ResponseActionResult,
  ResponseActionSpec,
} from "./provider-abi.js";

// ============================================================================
// Connector error class
// ============================================================================

/**
 * Throw from any of the lifecycle methods when the connector knows
 * exactly which failure category to surface. The runner maps this
 * to a :class:`ConnectorError` 1:1 (no loss of vendor metadata via
 * the optional ``vendorCode`` / ``vendorMessage`` fields).
 *
 * Common categories :
 *
 * - ``auth`` : credentials rejected / expired / missing
 * - ``not_found`` : target host / user / id doesn't exist upstream
 * - ``state_conflict`` : target is already in the desired state
 *   (e.g. host already isolated, user already disabled)
 * - ``transient`` : network blip, rate limit, upstream 5xx
 * - ``policy`` : the operation is forbidden by upstream policy
 *   (e.g. cannot disable the last admin, host has a maintenance hold)
 */
export class ConnectorAbiError extends Error {
  readonly category: FailureCategoryValue;
  readonly retryable: boolean;
  readonly vendorCode: string | null;
  readonly vendorMessage: string | null;

  constructor(args: {
    category: FailureCategoryValue;
    message: string;
    retryable?: boolean;
    vendorCode?: string | null;
    vendorMessage?: string | null;
  }) {
    super(args.message);
    this.name = "ConnectorAbiError";
    this.category = args.category;
    // Default : transient is retryable, everything else isn't.
    this.retryable = args.retryable ?? args.category === "transient";
    this.vendorCode = args.vendorCode ?? null;
    this.vendorMessage = args.vendorMessage ?? null;
  }

  /**
   * Convert to the wire-format :class:`ConnectorError` the runner
   * embeds in the audit chain. The runner usually calls this
   * implicitly ; expose it for connectors that want to surface
   * the error in a custom response body.
   */
  toConnectorError(): ConnectorError {
    return {
      specVersion: "1.0",
      category: this.category,
      message: this.message,
      retryable: this.retryable,
      vendorCode: this.vendorCode,
      vendorMessage: this.vendorMessage,
    };
  }
}

// ============================================================================
// AbiConnector base class
// ============================================================================

/**
 * Subclass this and implement the four lifecycle methods. The
 * subclass also exposes a static :attr:`capability` (an instance of
 * :class:`ConnectorCapability`) declaring which actions it supports
 * and how it authenticates ; the runner reads the capability to
 * route requests and decide whether dry-run is exercised.
 *
 * The constructor receives a free-form ``config`` dict (typically the
 * tenant's connector binding row). Subclasses should validate the
 * keys they care about in their constructor — fail fast on missing
 * required config.
 *
 * Example :
 *
 * ```ts
 * import {
 *   AbiConnector,
 *   ConnectorAbiError,
 *   type ConnectorCapability,
 *   type ResponseActionResult,
 *   type ResponseActionSpec,
 * } from "@warlog/spec";
 *
 * export class FalconConnector extends AbiConnector {
 *   static readonly capability: ConnectorCapability = ConnectorCapability.parse({ ... });
 *
 *   async authenticate(): Promise<void> {
 *     if (!this.config.apiKey) {
 *       throw new ConnectorAbiError({
 *         category: "auth", message: "Missing apiKey", retryable: false,
 *       });
 *     }
 *   }
 *
 *   async dryRun(spec: ResponseActionSpec): Promise<void> { ... }
 *
 *   async apply(spec: ResponseActionSpec): Promise<ResponseActionResult> { ... }
 *
 *   async verify(spec: ResponseActionSpec, result: ResponseActionResult): Promise<boolean> { ... }
 * }
 * ```
 */
export abstract class AbiConnector {
  /**
   * Static capability the runner reads to route + gate actions.
   * Subclasses MUST override with a concrete :class:`ConnectorCapability`.
   * Declared as ``unknown`` here so TypeScript doesn't require every
   * subclass to repeat the type ; the runner narrows at use time.
   */
  static readonly capability: ConnectorCapability;

  readonly config: Record<string, unknown>;

  constructor(config: Record<string, unknown> = {}) {
    this.config = config;
  }

  /**
   * Establish a session with the upstream vendor.
   *
   * The runner calls this exactly once at the start of every
   * action lifecycle, BEFORE dry-run. Failure here short-circuits
   * the whole lifecycle to FAILURE / category=auth — the audit
   * chain records a row at the APPLY phase with the auth error.
   *
   * Implementations SHOULD be idempotent : a second call within
   * the same lifecycle should be a no-op (or cached).
   */
  abstract authenticate(): Promise<void>;

  /**
   * Pre-flight check WITHOUT side effects upstream.
   *
   * Validates the target exists, the action is permitted under
   * current upstream state, and the connector can authenticate.
   * Throws :class:`ConnectorAbiError` with the appropriate
   * category when any precondition fails.
   *
   * The runner ALWAYS calls dryRun before apply (even when
   * ``spec.dryRun`` is false — dry-run is a precondition, not a
   * mode). Connectors that genuinely cannot validate without a
   * side effect SHOULD document the limitation in their
   * :attr:`capability.dryRun.scope`.
   */
  abstract dryRun(spec: ResponseActionSpec): Promise<void>;

  /**
   * Execute the action on the upstream vendor.
   *
   * MUST be idempotent w.r.t. ``spec.idempotencyKey`` : a re-invoke
   * with the same key MUST observe the prior state (no double-isolate,
   * no double-revoke). Vendor APIs that expose their own idempotency
   * token (CrowdStrike Falcon, SentinelOne) should forward
   * ``spec.idempotencyKey`` ; vendors without one need a tenant-side
   * dedup store.
   *
   * Returns a :class:`ResponseActionResult` with the outcome
  * (success / failure / expired) and optional vendor details.
  * ``pending_approval`` and ``denied`` are runner / approval-gate
  * outcomes used when execution stops before vendor mutation.
  * Throwing :class:`ConnectorAbiError` is equivalent to returning
  * a result with outcome=failure + the error's category.
   */
  abstract apply(
    spec: ResponseActionSpec,
  ): Promise<ResponseActionResult>;

  /**
   * Confirm the apply landed and the upstream state matches.
   *
   * Called by the runner AFTER apply, with the spec + apply result.
   * Returns ``true`` when verification succeeds, ``false`` when
   * the target state was not reached within the connector's
   * attempt budget. The runner records the verification outcome
   * as a separate audit row at the VERIFY phase.
   *
   * Connectors with no meaningful verify step (e.g. fire-and-
   * forget notifications) MAY return ``true`` unconditionally.
   */
  abstract verify(
    spec: ResponseActionSpec,
    result: ResponseActionResult,
  ): Promise<boolean>;
}

// ============================================================================
// AbiEnricher base class (read-side dual of AbiConnector)
// ============================================================================

/**
 * Optional base class for read-side connectors (VirusTotal,
 * AbuseIPDB, Shodan, GreyNoise, internal ML). Mirrors
 * :class:`warlog_spec.abi.AbiEnricher`. Implementations advertise
 * which artifact types they produce via
 * :attr:`ConnectorCapability.enrichment`.
 *
 * Distinct from :class:`AbiConnector` : no dry_run / approval /
 * verify lifecycle, no idempotency key. The enricher is read-only ;
 * the assessment IS the result.
 */
export abstract class AbiEnricher {
  static readonly capability: ConnectorCapability;
  readonly config: Record<string, unknown>;

  constructor(config: Record<string, unknown> = {}) {
    this.config = config;
  }

  abstract authenticate(): Promise<void>;

  /**
   * Produce a canonical artifact for the given subject + target.
   *
   * Returns ``null`` when the enricher had no information for the
   * subject. Returning a partial artifact is also acceptable when
   * the upstream had degraded data — the producer's confidence
   * band MUST reflect the degradation.
   *
   * The return type is intentionally ``unknown`` here ; concrete
   * subclasses narrow it to one of the canonical artifact types
   * (:class:`EnrichmentAssessment`, :class:`MitreAssessment`, …).
   * Untyped union avoids forcing every enricher to import every
   * artifact ; the runner re-validates via Zod at the call site.
   */
  abstract enrich(request: {
    subjectType: "alert" | "case";
    subjectId: string;
    target: unknown;
  }): Promise<unknown | null>;
}

File: packages/warlog-spec-py/src/warlog_spec/abi.py
"""ABI — abstract base classes a connector implements.

Two ABCs are exposed :

- :class:`AbiConnector` — write-side. Lifecycle hooks
  ``authenticate → dry_run → apply → verify``. Used by connectors
  that issue response actions against a vendor (host.isolate,
  user.disable, key.rotate, …).
- :class:`AbiEnricher` — read-side. Simpler lifecycle :
  ``authenticate → enrich``. Returns a canonical
  :class:`~warlog_spec.artifacts.EnrichmentAssessment` (or another
  envelope-bearing canonical shape). Used by connectors that PRODUCE
  context about an entity / IOC (VirusTotal, AbuseIPDB, Shodan, …).

A vendor that does both can implement both bases on a single class
via multiple inheritance, or ship two separate connector classes
sharing config. The runtime routes by the populated descriptors on
:class:`~warlog_spec.provider_abi.ConnectorCapability` —
``egress`` for writes, ``enrichment`` for reads.

Catches exceptions and maps them to :class:`ConnectorError` via the
failure model in :mod:`warlog_spec.provider_abi`. To shortcut
categorization, raise :class:`ConnectorAbiError` from connector code.

This module is the **public** integration surface : a third party
writes its connector by subclassing :class:`AbiConnector` or
:class:`AbiEnricher` from this package. The runtime that orchestrates
lifecycle, audit chain, approval gate, idempotency cache lives in
the Warlog backend and is NOT part of this package. Conformance only
requires you to implement the abstract methods correctly.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import ClassVar

from warlog_spec.artifacts import (
    CanonicalArtifact,
    EnrichmentRequest,
)
from warlog_spec.provider_abi import (
    ConnectorCapability,
    ConnectorError,
    FailureCategory,
    ResponseActionResult,
    ResponseActionSpec,
)


class ConnectorAbiError(Exception):
    """Categorized error a connector raises to short-circuit failure mapping.

    Use this when the connector knows the failure category at the
    call site (e.g. catching a vendor 401 → ``AUTH``). For unknown
    exceptions the runtime defaults to ``FailureCategory.TRANSIENT``
    (retryable).
    """

    def __init__(
        self,
        category: FailureCategory,
        message: str,
        *,
        retryable: bool | None = None,
        vendor_code: str | None = None,
        vendor_message: str | None = None,
    ) -> None:
        super().__init__(message)
        self.category = category
        self.message = message
        self.retryable = (
            retryable
            if retryable is not None
            else (category in {FailureCategory.TRANSIENT})
        )
        self.vendor_code = vendor_code
        self.vendor_message = vendor_message

    def to_connector_error(self) -> ConnectorError:
        return ConnectorError(
            category=self.category,
            message=self.message,
            retryable=self.retryable,
            vendor_code=self.vendor_code,
            vendor_message=self.vendor_message,
        )


class AbiConnector(ABC):
    """ABI-conformant connector base class.

    Subclasses must declare a class-level ``capability`` attribute
    and implement the four lifecycle hooks. The runtime calls them in
    order ``authenticate → dry_run → apply → verify``.

    ``apply`` MAY return a partial result if the upstream operation
    is asynchronous; ``verify`` is then expected to confirm the final
    state by polling the upstream vendor.
    """

    capability: ClassVar[ConnectorCapability]

    def __init__(self, config: dict[str, object]) -> None:
        self.config = config

    # -- Lifecycle hooks --------------------------------------------------

    @abstractmethod
    async def authenticate(self) -> None:
        """Acquire/refresh credentials.

        Raise :class:`ConnectorAbiError` with category ``AUTH`` on
        failure. Called once per execution by the runtime before
        ``dry_run``.
        """

    @abstractmethod
    async def dry_run(self, spec: ResponseActionSpec) -> None:
        """Validate the action without executing it.

        Connectors with ``capability.dry_run.supported = False`` MAY
        return immediately. Validation failures should raise
        :class:`ConnectorAbiError` (typically ``NOT_FOUND`` for missing
        subjects, ``STATE_CONFLICT`` for invalid current state,
        ``POLICY`` for vendor-side denials).
        """

    @abstractmethod
    async def apply(self, spec: ResponseActionSpec) -> ResponseActionResult:
        """Execute the action against the upstream vendor.

        .. important:: **Connector-side idempotency is REQUIRED.** The
           runtime's idempotency cache only protects against repeat
           calls within one process and one terminal outcome.
           Cross-process re-invokes after a ``PENDING_APPROVAL`` (a
           parallel approval landing during a re-poll) can both reach
           ``apply``. To prevent vendor-side double-execution, the
           connector MUST forward ``spec.idempotency_key`` to the
           upstream API as the vendor's idempotency token. Connectors
           that ignore the key are responsible for any duplicate side
           effects.

          Connector implementations should return ``SUCCESS``, ``FAILURE``,
          or ``EXPIRED``. ``PENDING_APPROVAL`` and ``DENIED`` are runner / approval
          gate outcomes used when execution stops before vendor mutation.
        """

    @abstractmethod
    async def verify(
        self,
        spec: ResponseActionSpec,
        result: ResponseActionResult,
    ) -> bool:
        """Poll the upstream vendor to confirm the action took effect.

        Return ``True`` once the target state is observed. The runtime
        retries on ``False`` with bounded backoff. Raise
        :class:`ConnectorAbiError` for hard failures.
        """


class AbiEnricher(ABC):
    """ABI-conformant enrichment connector base class.

    The read-side dual of :class:`AbiConnector`. Where a write-side
    connector mutates vendor state through a four-stage lifecycle,
    an enrichment connector PRODUCES a canonical artifact about a
    subject :

        ``authenticate → enrich(request) → CanonicalArtifact | None``

    No ``dry_run`` (reads are naturally non-destructive), no
    ``approval`` (the catalog defaults read-side actions to
    auto-execute), no ``verify`` (the assessment IS the result), no
    cross-process idempotency_key requirement (reads are naturally
    idempotent).

    Subclasses MUST :

    1. Declare a class-level :attr:`capability` whose
       ``enrichment`` :class:`~warlog_spec.provider_abi.EnrichmentDescriptor`
       declares what artifact types they produce (e.g.
       ``"enrichment.ioc_reputation"``, ``"mitre.assessment"``) and
       on what subjects (entity / IOC types).
    2. Implement :meth:`authenticate` (idempotent ; called once per
       runtime session).
    3. Implement :meth:`enrich` to return a
       :class:`~warlog_spec.artifacts.CanonicalArtifact` subclass
       (``EnrichmentAssessment``, ``MitreAssessment``, …) on a
       successful lookup, or ``None`` when the vendor has no data
       about the subject (clean miss, distinguished from an error).
       The connector's actual return type MUST be one of the shapes
       it advertised in ``capability.enrichment.produces_artifact_types``.

    Idiomatic shape : connector inspects the request's
    ``target`` type, queries the vendor API, maps the response into
    a typed payload, wraps it in an
    :class:`~warlog_spec.artifacts.ArtifactEnvelope` (copying
    ``subject_type`` / ``subject_id`` from the request so the
    artifact is self-attributing), and returns the composed
    :class:`~warlog_spec.artifacts.CanonicalArtifact`. Vendor-specific
    raw data is NOT returned ; the canonical envelope IS the contract.
    """

    capability: ClassVar[ConnectorCapability]

    def __init__(self, config: dict[str, object]) -> None:
        self.config = config

    @abstractmethod
    async def authenticate(self) -> None:
        """Acquire / refresh credentials.

        Raise :class:`ConnectorAbiError` with category ``AUTH`` on
        failure. Called once per session.
        """

    @abstractmethod
    async def enrich(
        self,
        request: EnrichmentRequest,
    ) -> CanonicalArtifact | None:
        """Produce a canonical artifact for ``request.target``.

        The returned artifact's envelope MUST carry
        ``subject_type`` / ``subject_id`` copied from the request —
        the connector is responsible for this attribution, not the
        runtime.

        Return ``None`` when the vendor has no data about the target
        (clean miss). Raise :class:`ConnectorAbiError` for hard
        failures :

        - ``AUTH`` for credential errors,
        - ``NOT_FOUND`` if the vendor explicitly says the target
          does not exist (distinct from a clean miss — only raise
          this when the vendor's API confirms non-existence rather
          than absence-of-data),
        - ``POLICY`` if the connector cannot enrich this target
          type (it should not have been routed here in the first
          place — the capability declares supports_*),
        - ``TRANSIENT`` for retryable failures (rate limits, network).
        """


__all__ = [
    "AbiConnector",
    "AbiEnricher",
    "ConnectorAbiError",
]
